authority/provisioner/webhook.go

```diff
@@ -0,0 +1,203 @@
+	Key             []byte          `json:"key"`
+	Serial          uint64          `json:"serial"`
+	CertType        uint32          `json:"certType"`
+	KeyID           string          `json:"keyID"`
+	ValidPrincipals []string        `json:"validPrincipals"`
+	ValidAfter      uint64          `json:"validAfter"`
+	ValidBefore     uint64          `json:"validBefore"`
+	Permissions     ssh.Permissions `json:"permissions"`
+	Reserved        []byte          `json:"reserved"`
+	SignatureKey    []byte          `json:"signatureKey"`
+	Signature       *ssh.Signature  `json:"Signature"`
+}
+
+type webhookRequestBody struct {
+	Timestamp time.Time `json:"timestamp"`
+	//nolint:revive // acronyms can use all caps
+	X509_CSR []byte `json:"csr,omitempty"`
+	//nolint:revive // acronyms can use all caps
+	SSH_CR         *sshRequest           `json:"ssh_cr,omitempty"`
+	Certificate    *x509util.Certificate `json:"certificate,omitempty"`
+	SSHCertificate *sshCert              `json:"ssh_certificate,omitempty"`
+}
+
+type WebhookResponseBody struct {
+	Data  map[string]interface{} `json:"data"`
+	Allow bool                   `json:"allow"`
+}
+
+func (w *Webhook) Do(ctx context.Context, client *http.Client, certReq interface{}, data map[string]interface{}) (*WebhookResponseBody, error) {
+	tmpl, err := template.New("url").Funcs(templates.StepFuncMap()).Parse(w.URL)
+	if err != nil {
+		return nil, err
+	}
+	buf := &bytes.Buffer{}
+	if err := tmpl.Execute(buf, data); err != nil {
+		return nil, err
+	}
+	url := buf.String()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	retries := 1
+retry:
+	reqBody := &webhookRequestBody{
+		Timestamp: time.Now(),
+	}
+	switch r := certReq.(type) {
+	case *x509.CertificateRequest:
+		if r != nil {
+			reqBody.X509_CSR = r.Raw
+		}
+	case sshutil.CertificateRequest:
+		reqBody.SSH_CR = &sshRequest{
+			Type:       r.Type,
+			KeyID:      r.KeyID,
+			Principals: r.Principals,
+		}
+		if r.Key != nil {
+			reqBody.SSH_CR.Key = r.Key.Marshal()
+		}
+	case *x509.Certificate:
+		reqBody.Certificate = x509util.FromX509Certificate(r)
+	case *ssh.Certificate:
+		reqBody.SSHCertificate = &sshCert{
+			Nonce:           r.Nonce,
+			Key:             r.Key.Marshal(),
+			Serial:          r.Serial,
+			CertType:        r.CertType,
+			KeyID:           r.KeyId,
+			ValidPrincipals: r.ValidPrincipals,
+			ValidAfter:      r.ValidAfter,
+			ValidBefore:     r.ValidBefore,
+			Permissions:     r.Permissions,
+			Reserved:        r.Reserved,
+		}
+		if r.Key != nil {
+			reqBody.SSHCertificate.Key = r.Key.Marshal()
+		}
+		if r.SignatureKey != nil {
+			reqBody.SSHCertificate.SignatureKey = r.SignatureKey.Marshal()
+		}
+	}
+
+	reqBytes, err := json.Marshal(reqBody)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(reqBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	secret, err := base64.StdEncoding.DecodeString(w.SigningSecret)
```

Comment:
The name `SigningSecret` is weird in this context. Just `Secret` or if we want to be verbose `MACSecret` would make more sense.

Reply:
acf0965c